drivers/timer/mchp_xec_rtos_timer.c

```diff
@@ -0,0 +1,271 @@
+ * thus reflects how much time we can reliably see expire between
+ * calls to elapsed() to read the COUNTFLAG bit.  So it needs to be
+ * set to be larger than the maximum time the interrupt might be
+ * masked.  Choosing a fraction of a tick is probably a good enough
+ * default, with an absolute minimum of 1k cyc.
+ */
+#define MIN_DELAY MAX(32, (CYC_PER_TICK/16))
+
+#define TICKLESS (IS_ENABLED(CONFIG_TICKLESS_KERNEL) &&			\
+		  !IS_ENABLED(CONFIG_QEMU_TICKLESS_WORKAROUND))
+
+/* VAL value above which we assume that a subsequent COUNTFLAG
+ * overflow seen in CTRL is real and not an artifact of wraparound
+ * timing.
+ */
+#define VAL_ABOUT_TO_WRAP 1
+
+static struct k_spinlock lock;
+
+static volatile u32_t last_load;
+
+static volatile u32_t cycle_count;
+
+static volatile u32_t announced_cycles;
+
+static volatile u32_t elapsed_counts;
+
+static ALWAYS_INLINE void timer_clear_intr(void)
+{
+	GIRQ23_REGS->SRC = MCHP_RTMR_GIRQ_VAL;
+	GIRQ23_REGS->RESULT = GIRQ23_REGS->SRC;
+	NVIC_ClearPendingIRQ(RTMR_IRQn);
+}
+
+static ALWAYS_INLINE void timer_preload_set(u32_t val)
+{
+	RTMR_REGS->PRLD = val & COUNTER_MAX;
+}
+
+static ALWAYS_INLINE void timer_counter_set(u32_t val)
+{
+	RTMR_REGS->CNT = val & COUNTER_MAX;
+}
+
+static ALWAYS_INLINE void timer_start(void)
+{
+	RTMR_REGS->CTRL = MCHP_RTMR_CTRL_BLK_EN + MCHP_RTMR_CTRL_HW_HALT_EN
+			  + MCHP_RTMR_CTRL_START;
+}
+
+/*
+ * Side-effect all RTOS timer registers are cleared.
+ */
+static ALWAYS_INLINE void timer_stop(void)
+{
+	RTMR_REGS->CTRL = 0;
+}
+
+static ALWAYS_INLINE void timer_pause(void)
+{
+	RTMR_REGS->CTRL |= MCHP_RTMR_CTRL_FW_HALT_EN;
+}
+
+static ALWAYS_INLINE void timer_unpause(void)
+{
+	RTMR_REGS->CTRL &= ~(MCHP_RTMR_CTRL_FW_HALT_EN);
+}
+
+static INLINE u32_t timer_is_paused(void)
+{
+	return (RTMR_REGS->CTRL & MCHP_RTMR_CTRL_FW_HALT_EN);
+}
+
+static void timer_ien(void)
+{
+	GIRQ23_REGS->EN_SET = MCHP_RTMR_GIRQ_VAL;
+}
+
+/* SYSTICK elapsed function */
+#if 0
+static u32_t elapsed(void)
+{
+	u32_t val, ctrl1, ctrl2;
+
+	ctrl1 = GIRQ23_REGS->SRC;
+	val = RTMR_REGS->CNT & COUNTER_MAX;
+	ctrl2 = GIRQ23_REGS->SRC;
+
+	overflow_cyc += (ctrl1 & MCHP_RTMR_GIRQ_VAL) ? last_load : 0;
+	if (val > VAL_ABOUT_TO_WRAP) {
+		int wrap = ctrl2 & MCHP_RTMR_GIRQ_VAL;
+
+		overflow_cyc += (wrap != 0) ? last_load : 0;
+	}
+
+	return (last_load - val) + overflow_cyc;
+}
+#endif
+
+/*
+ * RTOS Timer runs at 32KHz and counts down to 0. The number of counts
+ * elpapsed since the timer was started is Preload register - Count register.
+ * This routine is expected to return the number of ticks elapsed since the
+ * last time this routine was called. This implementation depends upon
+ * one-shot behaviour (auto-reload NOT enabled).
+ */
+static u32_t elapsed(void)
+{
+	u32_t tdiff = (last_load - (RTMR_REGS->CNT & COUNTER_MAX));
+	u32_t elapsed = 0;
+
+	if (tdiff != elapsed_counts) {
+		elapsed_counts = tdiff;
+		elapsed = tdiff;
+	}
+	return elapsed;
```

Comment:
This implementation looks wrong to me.  The return value of z_clock_elapsed()  (for which this is the basic engine) is the number of ticks that have elapsed since the last call to z_clock_announce().

This will do the right thing once.  Then it will start returning zeros (!) until the next cycle ticks over, at which point it will return the correct value.  Once.  Then it will break until the next cycle, etc...

If you need those semantics elsewhere in the code, you may need to split out two different elapsed() variants.